File: crates/kassi-signer/src/infisical.rs
use std::time::{Duration, Instant};

use base64::{engine::general_purpose::STANDARD as BASE64, Engine};
use reqwest::Client;
use secrecy::{ExposeSecret, SecretString};
use serde::{Deserialize, Serialize};
use tokio::sync::RwLock;

use crate::{Kms, SignerError};

/// Buffer before actual expiry to avoid using a token that's about to die.
const EXPIRY_BUFFER: Duration = Duration::from_secs(30);

#[derive(Debug, thiserror::Error)]
pub enum KmsError {
    #[error("http error: {0}")]
    Http(#[from] reqwest::Error),

    #[error("{operation} failed: {body}")]
    Api {
        operation: &'static str,
        body: String,
    },
}

struct TokenState {
    access_token: SecretString,
    expires_at: Instant,
}

const BASE_URL: &str = "https://app.infisical.com";

pub struct InfisicalKms {
    client: Client,
    client_id: SecretString,
    client_secret: SecretString,
    project_id: String,
    token: RwLock<TokenState>,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct LoginRequest<'a> {
    client_id: &'a str,
    client_secret: &'a str,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct LoginResponse {
    access_token: String,
    expires_in: u64,
}

#[derive(Deserialize)]
struct KeyResponse {
    key: KeyData,
}

#[derive(Deserialize)]
struct KeyData {
    id: String,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct CreateKeyRequest<'a> {
    project_id: &'a str,
    name: &'a str,
    encryption_algorithm: &'a str,
}

#[derive(Serialize)]
struct EncryptRequest<'a> {
    plaintext: &'a str,
}

#[derive(Deserialize)]
struct EncryptResponse {
    ciphertext: String,
}

#[derive(Serialize)]
struct DecryptRequest<'a> {
    ciphertext: &'a str,
}

#[derive(Deserialize)]
struct DecryptResponse {
    plaintext: String,
}

impl InfisicalKms {
    /// Authenticate with Infisical using universal auth (client ID + secret)
    /// and return an authenticated `InfisicalKms` instance.
    ///
    /// # Panics
    /// Panics if the HTTP client cannot be built (TLS backend unavailable).
    ///
    /// # Errors
    /// Returns `SignerError::Kms` if authentication fails.
    pub async fn login(
        client_id: impl Into<String>,
        client_secret: impl Into<String>,
        project_id: impl Into<String>,
    ) -> Result<Self, SignerError> {
        let client_id = SecretString::from(client_id.into());
        let client_secret = SecretString::from(client_secret.into());

        let client = Client::builder()
            .user_agent(format!(
                "kassi/{} ({})",
                env!("CARGO_PKG_VERSION"),
                option_env!("KASSI_GIT_SHA").unwrap_or("dev"),
            ))
            .timeout(Duration::from_secs(30))
            .connect_timeout(Duration::from_secs(5))
            .pool_max_idle_per_host(10)
            .build()
            .expect("failed to build HTTP client");

        let token_state = authenticate(
            &client,
            client_id.expose_secret(),
            client_secret.expose_secret(),
        )
        .await?;

        Ok(Self {
            client,
            client_id,
            client_secret,
            project_id: project_id.into(),
            token: RwLock::new(token_state),
        })
    }

    async fn get_key_id(&self, name: &str) -> Result<String, KmsError> {
        let token = self.valid_token().await?;

        let resp = self
            .client
            .get(format!("{BASE_URL}/api/v1/kms/keys/key-name/{name}"))
            .bearer_auth(token.expose_secret())
            .query(&[("projectId", &self.project_id)])
            .send()
            .await
            .map_err(KmsError::from)?;

        let resp = check_response(resp, "get key by name").await?;
        Ok(resp
            .json::<KeyResponse>()
            .await
            .map_err(KmsError::from)?
            .key
            .id)
    }

    /// Return a valid access token, refreshing if expired or about to expire.
    async fn valid_token(&self) -> Result<SecretString, KmsError> {
        {
            let state = self.token.read().await;
            if Instant::now() < state.expires_at {
                return Ok(state.access_token.clone());
            }
        }

        let mut state = self.token.write().await;
        // double-check: another task may have refreshed while we waited for the write lock
        if Instant::now() < state.expires_at {
            return Ok(state.access_token.clone());
        }

        tracing::info!("refreshing infisical access token");
        *state = authenticate(
            &self.client,
            self.client_id.expose_secret(),
            self.client_secret.expose_secret(),
        )
        .await?;

        Ok(state.access_token.clone())
    }
}

impl Kms for InfisicalKms {
    async fn create_key(&self, name: &str) -> Result<(), KmsError> {
        let token = self.valid_token().await?;

        let resp = self
            .client
            .post(format!("{BASE_URL}/api/v1/kms/keys"))
            .bearer_auth(token.expose_secret())
            .json(&CreateKeyRequest {
                project_id: &self.project_id,
                name,
                encryption_algorithm: "aes-256-gcm",
            })
            .send()
            .await
            .map_err(KmsError::from)?;

        check_response(resp, "create key").await?;
        Ok(())
    }

    async fn encrypt(&self, name: &str, plaintext: &[u8]) -> Result<String, KmsError> {
        let key_id = self.get_key_id(name).await?;
        let token = self.valid_token().await?;
        let b64 = BASE64.encode(plaintext);

        let resp = self
            .client
            .post(format!("{BASE_URL}/api/v1/kms/keys/{key_id}/encrypt"))
            .bearer_auth(token.expose_secret())
            .json(&EncryptRequest { plaintext: &b64 })
            .send()
            .await
            .map_err(KmsError::from)?;

        let resp = check_response(resp, "encrypt").await?;
        Ok(resp
            .json::<EncryptResponse>()
            .await
            .map_err(KmsError::from)?
            .ciphertext)
    }

    async fn decrypt(&self, name: &str, ciphertext: &str) -> Result<Vec<u8>, KmsError> {
        let key_id = self.get_key_id(name).await?;
        let token = self.valid_token().await?;

        let resp = self
            .client
            .post(format!("{BASE_URL}/api/v1/kms/keys/{key_id}/decrypt"))
            .bearer_auth(token.expose_secret())
            .json(&DecryptRequest { ciphertext })
            .send()
            .await
            .map_err(KmsError::from)?;

        let resp = check_response(resp, "decrypt").await?;
        let plaintext_b64 = resp
            .json::<DecryptResponse>()
            .await
            .map_err(KmsError::from)?
            .plaintext;

        BASE64.decode(&plaintext_b64).map_err(|e| KmsError::Api {
            operation: "decrypt (base64 decode)",
            body: e.to_string(),
        })
    }
}

async fn authenticate(
    client: &Client,
    client_id: &str,
    client_secret: &str,
) -> Result<TokenState, KmsError> {
    let resp = client
        .post(format!("{BASE_URL}/api/v1/auth/universal-auth/login"))
        .json(&LoginRequest {
            client_id,
            client_secret,
        })
        .send()
        .await?;

    let resp = check_response(resp, "login").await?;
    let login = resp.json::<LoginResponse>().await?;

    Ok(TokenState {
        access_token: SecretString::from(login.access_token),
        expires_at: (Instant::now() + Duration::from_secs(login.expires_in))
            .checked_sub(EXPIRY_BUFFER)
            .unwrap_or_else(Instant::now),
    })
}

async fn check_response(
    resp: reqwest::Response,
    operation: &'static str,
) -> Result<reqwest::Response, KmsError> {
    if !resp.status().is_success() {
        return Err(KmsError::Api {
            operation,
            body: resp.text().await.unwrap_or_default(),
        });
    }
    Ok(resp)
}

#[cfg(test)]
mod tests {
    use super::*;

    async fn dev_client() -> InfisicalKms {
        InfisicalKms::login("test-client-id", "test-client-secret", "test-project-id")
            .await
            .unwrap()
    }

    #[tokio::test]
    #[ignore = "requires infisical dev server"]
    async fn encrypt_decrypt_round_trip() {
        let kms = dev_client().await;
        let key_name = "test-round-trip";

        kms.create_key(key_name).await.unwrap();

        let plaintext = b"super secret merchant seed data!";
        let ciphertext = kms.encrypt(key_name, plaintext).await.unwrap();

        let decrypted = kms.decrypt(key_name, &ciphertext).await.unwrap();
        assert_eq!(decrypted, plaintext);
    }

    #[tokio::test]
    #[ignore = "requires infisical dev server"]
    async fn decrypt_with_wrong_key_fails() {
        let kms = dev_client().await;

        kms.create_key("key-a").await.unwrap();
        kms.create_key("key-b").await.unwrap();

        let ciphertext = kms.encrypt("key-a", b"secret").await.unwrap();
        let result = kms.decrypt("key-b", &ciphertext).await;
        assert!(result.is_err());
    }
}

File: internal/signer/infisical.go
package signer

import (
	"context"
	"encoding/base64"
	"fmt"
	"sync"
	"sync/atomic"
	"time"

	"github.com/imroc/req/v3"
	"golang.org/x/sync/singleflight"
)

type cachedToken struct {
	value     string
	expiresAt time.Time
}

type InfisicalKMS struct {
	clientID     string
	clientSecret string
	projectID    string
	client       *req.Client
	token        atomic.Value   // *cachedToken
	tokenFlight  singleflight.Group
	keys         sync.Map       // name -> keyId (string)
}

func NewInfisicalKMS(clientID, clientSecret, projectID string) *InfisicalKMS {
	kms := &InfisicalKMS{
		clientID:     clientID,
		clientSecret: clientSecret,
		projectID:    projectID,
	}

	kms.client = req.C().
		SetBaseURL("https://app.infisical.com/api").
		SetTimeout(10 * time.Second).
		SetUserAgent("kassi").
		OnBeforeRequest(kms.ensureAuth)

	return kms
}

func (k *InfisicalKMS) ensureAuth(_ *req.Client, r *req.Request) error {
	if r.RawURL == "/v1/auth/universal-auth/login" {
		return nil
	}

	if v, ok := k.token.Load().(*cachedToken); ok && time.Now().Before(v.expiresAt) {
		r.SetBearerAuthToken(v.value)
		return nil
	}

	result, err, _ := k.tokenFlight.Do("token", func() (any, error) {
		if v, ok := k.token.Load().(*cachedToken); ok && time.Now().Before(v.expiresAt) {
			return v.value, nil
		}

		var authResp struct {
			AccessToken string `json:"accessToken"` //nolint:gosec
			ExpiresIn   int64  `json:"expiresIn"`
		}

		resp, err := k.client.R().
			SetContext(r.Context()).
			SetBodyJsonMarshal(map[string]string{
				"clientId":     k.clientID,
				"clientSecret": k.clientSecret,
			}).
			SetSuccessResult(&authResp).
			Post("/v1/auth/universal-auth/login")
		if err != nil {
			return nil, fmt.Errorf("authenticating with infisical: %w", err)
		}
		if !resp.IsSuccessState() {
			return nil, fmt.Errorf("infisical auth failed (status %d): %s", resp.StatusCode, resp.String())
		}

		k.token.Store(&cachedToken{
			value:     authResp.AccessToken,
			expiresAt: time.Now().Add(time.Duration(authResp.ExpiresIn)*time.Second - 30*time.Second),
		})

		return authResp.AccessToken, nil
	})
	if err != nil {
		return err
	}

	r.SetBearerAuthToken(result.(string))
	return nil
}

func (k *InfisicalKMS) resolveKeyID(ctx context.Context, name string) (string, error) {
	if id, ok := k.keys.Load(name); ok {
		return id.(string), nil
	}

	var result struct {
		Key struct {
			ID string `json:"id"`
		} `json:"key"`
	}

	resp, err := k.client.R().
		SetContext(ctx).
		SetPathParam("keyName", name).
		SetQueryParam("projectId", k.projectID).
		SetSuccessResult(&result).
		Get("/v1/kms/keys/key-name/{keyName}")
	if err != nil {
		return "", fmt.Errorf("resolving KMS key %s: %w", name, err)
	}
	if !resp.IsSuccessState() {
		return "", fmt.Errorf("infisical get key failed (status %d): %s", resp.StatusCode, resp.String())
	}

	k.keys.Store(name, result.Key.ID)
	return result.Key.ID, nil
}

func (k *InfisicalKMS) CreateKey(ctx context.Context, name string) error {
	var result struct {
		Key struct {
			ID string `json:"id"`
		} `json:"key"`
	}

	resp, err := k.client.R().
		SetContext(ctx).
		SetBodyJsonMarshal(map[string]string{
			"projectId":           k.projectID,
			"name":                name,
			"encryptionAlgorithm": "aes-256-gcm",
		}).
		SetSuccessResult(&result).
		Post("/v1/kms/keys")
	if err != nil {
		return fmt.Errorf("creating KMS key: %w", err)
	}
	if !resp.IsSuccessState() {
		return fmt.Errorf("infisical create key failed (status %d): %s", resp.StatusCode, resp.String())
	}

	k.keys.Store(name, result.Key.ID)
	return nil
}

func (k *InfisicalKMS) Encrypt(ctx context.Context, name string, plaintext []byte) (string, error) {
	keyID, err := k.resolveKeyID(ctx, name)
	if err != nil {
		return "", err
	}

	var result struct {
		Ciphertext string `json:"ciphertext"`
	}

	resp, err := k.client.R().
		SetContext(ctx).
		SetPathParam("keyId", keyID).
		SetBodyJsonMarshal(map[string]string{
			"plaintext": base64.StdEncoding.EncodeToString(plaintext),
		}).
		SetSuccessResult(&result).
		Post("/v1/kms/keys/{keyId}/encrypt")
	if err != nil {
		return "", fmt.Errorf("encrypting via KMS: %w", err)
	}
	if !resp.IsSuccessState() {
		return "", fmt.Errorf("infisical encrypt failed (status %d): %s", resp.StatusCode, resp.String())
	}

	return result.Ciphertext, nil
}

func (k *InfisicalKMS) Decrypt(ctx context.Context, name string, ciphertext string) ([]byte, error) {
	keyID, err := k.resolveKeyID(ctx, name)
	if err != nil {
		return nil, err
	}

	var result struct {
		Plaintext string `json:"plaintext"`
	}

	resp, err := k.client.R().
		SetContext(ctx).
		SetPathParam("keyId", keyID).
		SetBodyJsonMarshal(map[string]string{"ciphertext": ciphertext}).
		SetSuccessResult(&result).
		Post("/v1/kms/keys/{keyId}/decrypt")
	if err != nil {
		return nil, fmt.Errorf("decrypting via KMS: %w", err)
	}
	if !resp.IsSuccessState() {
		return nil, fmt.Errorf("infisical decrypt failed (status %d): %s", resp.StatusCode, resp.String())
	}

	decoded, err := base64.StdEncoding.DecodeString(result.Plaintext)
	if err != nil {
		return nil, fmt.Errorf("decoding plaintext: %w", err)
	}

	return decoded, nil
}
